keps/sig-api-machinery/NNNN-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,742 @@
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes clusters architecture have a "hub-and-spoke" pattern and all the components are required to communicate
+with the API servers. Kubernetes components use client-go for these communications.
+
+Creating high available clusters requires to use multiple control-plane nodes, however, clients need to use `Services` abstractions or external load balancers for bootstrapping and/or HA.
+
+This proposal outlines a new feature in client-go that would allow clients to not requires external load balancers or `Services` for bootstrapping and/or HA.
+
+## Motivation
+
+Currently, when deploying high available clusters, clients can only benefit of it by using external load balancers or mechanisms that balance the connection to the corresponding control-plane.
+This also has the "chicken and egg" problem on some components like kube-proxy and kubenet and "in-cluster" clients, that require `Services` to load balance the connection.
+
+Ref: https://github.com/kubernetes/kubernetes/issues/18174
+
+### Goals
+
+- Allow users to don't require external load balancers between cluster and control-plane
+- Allow components to bootstrap without any dependency on external load balancers or `Services` implementations
+- Increase the resilience of the cluster
+### Non-Goals
+
+- The goal is to offer the opt-in functionality, any component change to use it is out of scope.
+- Provide a lot of nines of availability
+- Work with clients outside of the cluster network
```

Comment:
As ability to connect to multiple API servers may suggest this as well.
```suggestion
- Work with clients outside of the cluster network
- Provide a client-side load balancing solution to distribute the load between the available API servers
```

Reply:
good point